imagersite/imager_images/views.py

```diff
@@ -1,3 +1,140 @@
-from django.shortcuts import render
+from __future__ import unicode_literals
+from __future__ import division
+from django.views.generic import DetailView
+from django.views.generic.edit import CreateView, UpdateView
+from imager_images.models import Photo, Album, Face
+from django.db.models import Q
+from django.http import Http404
+import os
 
-# Create your views here.
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+def get_faces(photo):
+    import Algorithmia
+    import base64
+    Algorithmia.apiKey = os.environ['ALGORITHMIA_KEY']
+    file_path = 'media/' + str(photo.file)
+    file_path = os.path.join(BASE_DIR, file_path)
+
+    with open(file_path, 'rb') as img:
+        b64 = base64.b64encode(img.read())
+
+    result = Algorithmia.algo("/ANaimi/FaceDetection").pipe(b64)
+
+    faces = []
+    for rect in result:
+        face = Face()
+        face.photo = photo
+        face.name = '?'
+        face.x = rect['x']
+        face.y = rect['y']
+        face.width = rect['width']
+        face.height = rect['height']
+        face.save()
+        faces.append(face)
+    return faces
+
+
+class PhotoView(DetailView):
+    model = Photo
+    template_name = 'photo.html'
+    detect = False
+
+    def get_queryset(self, *args, **kwargs):
+        return super(PhotoView, self).get_queryset(*args, **kwargs).filter(
+            Q(user=self.request.user) | Q(published='Public'))
+
+    def get_context_data(self, **kwargs):
+        context = super(DetailView, self).get_context_data(**kwargs)
+        if self.detect and len(self.object.faces.all()) == 0:
+            get_faces(self.object)
+
+        context['faces'] = self.object.faces.all()
```

Comment:
if you think about it, you've got self.object in your template, which means you already have access to the faces without needing to put them into context separately.